cpp/doxygen/developer_guide/DOCUMENTATION.md

```diff
@@ -0,0 +1,530 @@
+them this way in its output.
+
+The doxygen output includes a _Modules_ page that organizes items into groups specified using the
+[Grouping doxygen commands](https://www.doxygen.nl/manual/grouping.html). These commands can group
+common functions across header files, source files, and even namespaces. Groups can also be nested
+by defining new groups within existing groups.
+
+For libcuspatial, all the group hierarchy is defined in the 
+[doxygen_groups.h](../include/doxygen_groups.h) header file. The `doxygen_groups.h` file does not
+need to be included in any other source file, because the definitions in this file are used only by
+the doxygen tool to generate groups in the _Modules_ page. Modify this file only to add or update
+groups. The existing groups have been carefully structured and named, so new groups should be added
+thoughtfully.
+
+When creating a new API, specify its group using the
+[\@ingroup](https://www.doxygen.nl/manual/commands.html#cmdingroup) tag and the group reference id
+from the [doxygen_groups.h](../include/doxygen_groups.h) file.
+
+    namespace cuspatial {
+
+    /**
+     * @brief ...
+     *
+     * @ingroup distance
+     *
+     * @tparam...
+     * @param ...
+     * @return ...
+     */
+    template <class Cart2dItA, class Cart2dItB, class OutputIt>
+    OutputIt pairwise_point_distance(Cart2dItA points1_first, ...);
+
+    }  // namespace cuspatial
+
+You can also use the \@addtogroup with a `@{ ... @}` pair to automatically include doxygen comment
+blocks as part of a group.
+
+    namespace cuspatial {
+    /**
+     * @addtogroup coordinate_transform
+     * @{
+     */
+
+    /**
+     * @brief ...
+     *
+     * @param ...
+     * @return ...
+     */
+    template <class Cart2dItA, class Cart2dItB, class OutputIt>
+    OutputIt pairwise_point_distance(Cart2dItA points1_first, ...);
+
+    /** @} */
+    }  // namespace cuspatial
+
+This just saves adding \@ingroup to individual doxygen comment blocks within a file. Make sure a
+blank line is included after the \@addtogroup command block so doxygen knows it does not apply to
+whatever follows in the source code. Note that doxygen will not assign groups to items if the
+\@addtogroup with `@{ ... @}` pair includes a namespace declaration. So include the \@addtogroup
+and `@{ ... @}` between the namespace declaration braces as shown in the example above.
+
+Summary of groups tags
+| Tag/Command | Where to use |
+| ----------- | ------------ |
+| \@defgroup | For use only in [doxygen_groups.h](../include/doxygen_groups.h) and should include the group's title. |
+| \@ingroup | Use inside individual doxygen block comments for declaration statements in a header file. |
+| \@addtogroup | Use instead of \@ingroup for multiple declarations in the same file within a namespace declaration. Do not specify a group title. |
+| `@{ ... @}` |  Use only with \@addtogroup. |
```

Comment:
I think a list of the groups in `doxygen_groups.h` and their purpose might be appropriate here. Or a link to it telling the user to go read them.

Reply:
Done.